### PR TITLE
[Group 4] Enable nullable annotations for `Microsoft.Extensions.Logging.TraceSource`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to https://github.com/dotnet/runtime/issues/43605

Questions:
- `BeginScope` should be annotated with `where TState : notnull`, but it cannot, since it's not annotated this way in `ILogger`
https://github.com/dotnet/runtime/blob/3e2a45c1ca2c990e9fad9fff4e3a6230d2dfc9a2/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs#L73
- `Microsoft.Extensions.Logging.EventLog` cannot be annotated since `System.Diagnostics.EventLog` is not annotated yet. Should something be done about it, or let it be for now?

### CODE REVIEW

**Comment by eerhardt** [comment]
> BeginScope should be annotated with where TState : notnull, but it cannot, since it's not annotated this way in ILogger

We can still change it in ILogger, if that is the correct thing to do. Doing a real quick look, I think it is.

BeginScope should also be annotated as returning `IDisposable?` because there are multiple places that return `null`:

https://github.com/dotnet/runtime/blob/eb51b02b158c3ff71a1ec7eac8a211d1d464c1a5/src/libraries/Microsoft.Extensions.Logging.EventLog/src/EventLogLogger.cs#L51-L54

https://github.com/dotnet/aspnetcore/blob/17695facaeaf9d6d83aea9bfe87747180922117a/src/Logging.AzureAppServices/src/BatchingLogger.cs#L21-L24

**Comment by eerhardt** [comment]
> Microsoft.Extensions.Logging.EventLog cannot be annotated since System.Diagnostics.EventLog is not annotated yet. Should something be done about it, or let it be for now?

@maxkoshevoi - is annotating System.Diagnostics.EventLog something you would be interested in doing as well? If not, that is totally fine.

File: src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs [hunk 1]
```diff
@@ -70,9 +70,9 @@ private static TraceEventType GetEventType(LogLevel logLevel)
             }
         }
 
-        public IDisposable BeginScope<TState>(TState state)
+        public IDisposable BeginScope<TState>(TState state) where TState : notnull
         {
-            return new TraceSourceScope(state);
+            return new TraceSourceScope(state!);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs:75]:
```suggestion
            return new TraceSourceScope(state);
```

File: src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ public TraceSourceLogger(DiagnosticsTraceSource traceSource)
             _traceSource = traceSource;
         }
 
-        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string>? formatter)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs:19]: I think this should match the ILogger signature.

```suggestion
        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
```

**Review by eerhardt — APPROVED** [review]
LGTM. Just two tiny comments.

**Comment by maxkoshevoi** [comment]
> is annotating System.Diagnostics.EventLog something you would be interested in doing as well? If not, that is totally fine.

Sure, I was running out of things to annotate anyway 😄
Does this apply to `System.Security.Cryptography.Xml` too?

**Comment by eerhardt** [comment]
> Does this apply to System.Security.Cryptography.Xml too?

For `Microsoft.Extensions.Configuration.Xml`, I only know of 1 place that uses `System.Security.Cryptography.Xml`:

https://github.com/dotnet/runtime/blob/000e1a8409ebd2dcc9539a20c668dd8c1c99627f/src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlDocumentDecryptor.cs#L92-L102

I think it would be pretty easy to nullable annotate `Microsoft.Extensions.Configuration.Xml` without having `System.Security.Cryptography.Xml` annotated.

But if you wanted to annotate `System.Security.Cryptography.Xml`, we should check with @bartonjs before starting on it, to be sure he would be open to accepting a PR to nullable annotate it.

**Review by maryamariyan — APPROVED** [review]

**Comment by eerhardt** [comment]
@maryamariyan - did the CI run for this PR?

**Review by maryamariyan — COMMENTED** [review]
LGTM. @eerhardt not sure what happened with the CI here.

**Pull request merged**